// xero-node/dist/gen/model/payroll-uk/nICategoryOneOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NICategoryOneOf = void 0;
class NICategoryOneOf {
    static getAttributeTypeMap() {
        return NICategoryOneOf.attributeTypeMap;
    }
}
exports.NICategoryOneOf = NICategoryOneOf;
NICategoryOneOf.discriminator = undefined;
NICategoryOneOf.attributeTypeMap = [
    {
        "name": "niCategory",
        "baseName": "niCategory",
        "type": "NICategoryOneOf.NiCategoryEnum"
    }
];
(function (NICategoryOneOf) {
    let NiCategoryEnum;
    (function (NiCategoryEnum) {
        NiCategoryEnum[NiCategoryEnum["F"] = 'F'] = "F";
        NiCategoryEnum[NiCategoryEnum["I"] = 'I'] = "I";
        NiCategoryEnum[NiCategoryEnum["L"] = 'L'] = "L";
        NiCategoryEnum[NiCategoryEnum["S"] = 'S'] = "S";
        NiCategoryEnum[NiCategoryEnum["N"] = 'N'] = "N";
        NiCategoryEnum[NiCategoryEnum["E"] = 'E'] = "E";
        NiCategoryEnum[NiCategoryEnum["D"] = 'D'] = "D";
        NiCategoryEnum[NiCategoryEnum["K"] = 'K'] = "K";
    })(NiCategoryEnum = NICategoryOneOf.NiCategoryEnum || (NICategoryOneOf.NiCategoryEnum = {}));
})(NICategoryOneOf || (exports.NICategoryOneOf = NICategoryOneOf = {}));
