// xero-node/dist/gen/model/payroll-uk/nICategoryLetter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NICategoryLetter = void 0;
/**
* The employee\'s NI Category letter.
*/
var NICategoryLetter;
(function (NICategoryLetter) {
    NICategoryLetter[NICategoryLetter["A"] = 'A'] = "A";
    NICategoryLetter[NICategoryLetter["B"] = 'B'] = "B";
    NICategoryLetter[NICategoryLetter["C"] = 'C'] = "C";
    NICategoryLetter[NICategoryLetter["F"] = 'F'] = "F";
    NICategoryLetter[NICategoryLetter["H"] = 'H'] = "H";
    NICategoryLetter[NICategoryLetter["I"] = 'I'] = "I";
    NICategoryLetter[NICategoryLetter["J"] = 'J'] = "J";
    NICategoryLetter[NICategoryLetter["L"] = 'L'] = "L";
    NICategoryLetter[NICategoryLetter["M"] = 'M'] = "M";
    NICategoryLetter[NICategoryLetter["S"] = 'S'] = "S";
    NICategoryLetter[NICategoryLetter["V"] = 'V'] = "V";
    NICategoryLetter[NICategoryLetter["X"] = 'X'] = "X";
    NICategoryLetter[NICategoryLetter["Z"] = 'Z'] = "Z";
    NICategoryLetter[NICategoryLetter["D"] = 'D'] = "D";
    NICategoryLetter[NICategoryLetter["E"] = 'E'] = "E";
    NICategoryLetter[NICategoryLetter["K"] = 'K'] = "K";
    NICategoryLetter[NICategoryLetter["N"] = 'N'] = "N";
})(NICategoryLetter || (exports.NICategoryLetter = NICategoryLetter = {}));
