// xero-node/dist/gen/api/payrollAUV2Api.js
"use strict";
/**
 * Xero Payroll AU API 2.0
 * This is the Xero Payroll API 2.0 for orgs in Australia region.
 *
 * The version of the OpenAPI document: 11.1.0
 * Contact: api@xero.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PayrollAuV2Api = exports.PayrollAuV2ApiApiKeys = void 0;
const models_1 = require("../model/payroll-au-v2/models");
const ApiError_1 = require("../../model/ApiError");
const axios_1 = __importDefault(require("axios"));
const models_2 = require("../model/payroll-au-v2/models");
let defaultBasePath = 'https://api.xero.com/payroll.xro/2.0';
// ===============================================
// This file is autogenerated - Please do not edit
// ===============================================
var PayrollAuV2ApiApiKeys;
(function (PayrollAuV2ApiApiKeys) {
})(PayrollAuV2ApiApiKeys || (exports.PayrollAuV2ApiApiKeys = PayrollAuV2ApiApiKeys = {}));
class PayrollAuV2Api {
    constructor(basePathOrUsername, password, basePath) {
        this._basePath = defaultBasePath;
        this.defaultHeaders = { 'user-agent': 'xero-node-14.0.0' };
        this._useQuerystring = false;
        this.binaryHeaders = {};
        this.authentications = {
            'default': new models_1.VoidAuth(),
            'OAuth2': new models_2.OAuth(),
        };
        if (password) {
            if (basePath) {
                this.basePath = basePath;
            }
        }
        else {
            if (basePathOrUsername) {
                this.basePath = basePathOrUsername;
            }
        }
    }
    set useQuerystring(value) {
        this._useQuerystring = value;
    }
    set basePath(basePath) {
        this._basePath = basePath;
    }
    get basePath() {
        return this._basePath;
    }
    setDefaultAuthentication(auth) {
        this.authentications.default = auth;
    }
    setApiKey(key, value) {
        this.authentications[PayrollAuV2ApiApiKeys[key]].apiKey = value;
    }
    set accessToken(token) {
        this.authentications.OAuth2.accessToken = token;
    }
    /**
     *
     * @summary Approves a specific timesheet
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     * @param idempotencyKey This allows you to safely retry requests without the risk of duplicate processing. 128 character max.
     */
    approveTimesheet(xeroTenantId_1, timesheetID_1, idempotencyKey_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, idempotencyKey, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}/Approve'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling approveTimesheet.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling approveTimesheet.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Idempotency-Key'] = models_1.ObjectSerializer.serialize(idempotencyKey, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Creates a new timesheet
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheet
     * @param idempotencyKey This allows you to safely retry requests without the risk of duplicate processing. 128 character max.
     */
    createTimesheet(xeroTenantId_1, timesheet_1, idempotencyKey_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheet, idempotencyKey, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets';
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling createTimesheet.');
            }
            // verify required parameter 'timesheet' is not null or undefined
            if (timesheet === null || timesheet === undefined) {
                throw new Error('Required parameter timesheet was null or undefined when calling createTimesheet.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Idempotency-Key'] = models_1.ObjectSerializer.serialize(idempotencyKey, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: models_1.ObjectSerializer.serialize(timesheet, "Timesheet"),
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Creates a new timesheet line for a specific timesheet using a unique timesheet ID
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     * @param timesheetLine
     * @param idempotencyKey This allows you to safely retry requests without the risk of duplicate processing. 128 character max.
     */
    createTimesheetLine(xeroTenantId_1, timesheetID_1, timesheetLine_1, idempotencyKey_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, timesheetLine, idempotencyKey, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}/Lines'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling createTimesheetLine.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling createTimesheetLine.');
            }
            // verify required parameter 'timesheetLine' is not null or undefined
            if (timesheetLine === null || timesheetLine === undefined) {
                throw new Error('Required parameter timesheetLine was null or undefined when calling createTimesheetLine.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Idempotency-Key'] = models_1.ObjectSerializer.serialize(idempotencyKey, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: models_1.ObjectSerializer.serialize(timesheetLine, "TimesheetLine"),
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetLineObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Deletes a specific timesheet
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     */
    deleteTimesheet(xeroTenantId_1, timesheetID_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling deleteTimesheet.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling deleteTimesheet.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'DELETE',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetLine");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Deletes a specific timesheet line
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     * @param timesheetLineID Identifier for the timesheet line
     */
    deleteTimesheetLine(xeroTenantId_1, timesheetID_1, timesheetLineID_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, timesheetLineID, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}/Lines/{TimesheetLineID}'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)))
                .replace('{' + 'TimesheetLineID' + '}', encodeURIComponent(String(timesheetLineID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling deleteTimesheetLine.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling deleteTimesheetLine.');
            }
            // verify required parameter 'timesheetLineID' is not null or undefined
            if (timesheetLineID === null || timesheetLineID === undefined) {
                throw new Error('Required parameter timesheetLineID was null or undefined when calling deleteTimesheetLine.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'DELETE',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetLine");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Retrieve a specific timesheet by using a unique timesheet ID
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     */
    getTimesheet(xeroTenantId_1, timesheetID_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling getTimesheet.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling getTimesheet.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Retrieves timesheets
     * @param xeroTenantId Xero identifier for Tenant
     * @param page Page number which specifies the set of records to retrieve. By default the number of the records per set is 100.
     * @param filter Filter by employeeId and/or payrollCalendarId
     * @param status filter results by any timesheets with a matching timesheet status
     * @param startDate filter results by any timesheets with a startDate on or after the provided date
     * @param endDate filter results by any timesheets with a endDate on or before the provided date
     * @param sort sort the order of timesheets returned. The default is based on the timesheets createdDate, sorted oldest to newest. Currently, the only other option is to reverse the order based on the timesheets startDate, sorted newest to oldest.
     */
    getTimesheets(xeroTenantId_1, page_1, filter_1, status_1, startDate_1, endDate_1, sort_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, page, filter, status, startDate, endDate, sort, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets';
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling getTimesheets.');
            }
            if (page !== undefined) {
                localVarQueryParameters['page'] = models_1.ObjectSerializer.serialize(page, "number");
            }
            if (filter !== undefined) {
                localVarQueryParameters['filter'] = models_1.ObjectSerializer.serialize(filter, "string");
            }
            if (status !== undefined) {
                localVarQueryParameters['status'] = models_1.ObjectSerializer.serialize(status, "string");
            }
            if (startDate !== undefined) {
                localVarQueryParameters['startDate'] = models_1.ObjectSerializer.serialize(startDate, "string");
            }
            if (endDate !== undefined) {
                localVarQueryParameters['endDate'] = models_1.ObjectSerializer.serialize(endDate, "string");
            }
            if (sort !== undefined) {
                localVarQueryParameters['sort'] = models_1.ObjectSerializer.serialize(sort, "string");
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'GET',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "Timesheets");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Reverts a specific timesheet to draft
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     * @param idempotencyKey This allows you to safely retry requests without the risk of duplicate processing. 128 character max.
     */
    revertTimesheet(xeroTenantId_1, timesheetID_1, idempotencyKey_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, idempotencyKey, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}/RevertToDraft'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling revertTimesheet.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling revertTimesheet.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Idempotency-Key'] = models_1.ObjectSerializer.serialize(idempotencyKey, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'POST',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: {},
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
    /**
     *
     * @summary Updates a specific timesheet line for a specific timesheet
     * @param xeroTenantId Xero identifier for Tenant
     * @param timesheetID Identifier for the timesheet
     * @param timesheetLineID Identifier for the timesheet line
     * @param timesheetLine
     * @param idempotencyKey This allows you to safely retry requests without the risk of duplicate processing. 128 character max.
     */
    updateTimesheetLine(xeroTenantId_1, timesheetID_1, timesheetLineID_1, timesheetLine_1, idempotencyKey_1) {
        return __awaiter(this, arguments, void 0, function* (xeroTenantId, timesheetID, timesheetLineID, timesheetLine, idempotencyKey, options = { headers: {} }) {
            const localVarPath = this.basePath + '/Timesheets/{TimesheetID}/Lines/{TimesheetLineID}'
                .replace('{' + 'TimesheetID' + '}', encodeURIComponent(String(timesheetID)))
                .replace('{' + 'TimesheetLineID' + '}', encodeURIComponent(String(timesheetLineID)));
            let localVarQueryParameters = {};
            let localVarHeaderParams = Object.assign({}, this.defaultHeaders);
            let localVarFormParams = {};
            let acceptHeadersFromSpec = [
                "application/json"
            ];
            const isBufferType = acceptHeadersFromSpec.includes("application/pdf") || acceptHeadersFromSpec.includes("application/octet-stream") || acceptHeadersFromSpec.includes("application/jpg");
            const responseTypeOption = isBufferType ? "arraybuffer" : "json";
            // verify required parameter 'xeroTenantId' is not null or undefined
            if (xeroTenantId === null || xeroTenantId === undefined) {
                throw new Error('Required parameter xeroTenantId was null or undefined when calling updateTimesheetLine.');
            }
            // verify required parameter 'timesheetID' is not null or undefined
            if (timesheetID === null || timesheetID === undefined) {
                throw new Error('Required parameter timesheetID was null or undefined when calling updateTimesheetLine.');
            }
            // verify required parameter 'timesheetLineID' is not null or undefined
            if (timesheetLineID === null || timesheetLineID === undefined) {
                throw new Error('Required parameter timesheetLineID was null or undefined when calling updateTimesheetLine.');
            }
            // verify required parameter 'timesheetLine' is not null or undefined
            if (timesheetLine === null || timesheetLine === undefined) {
                throw new Error('Required parameter timesheetLine was null or undefined when calling updateTimesheetLine.');
            }
            localVarHeaderParams['Xero-Tenant-Id'] = models_1.ObjectSerializer.serialize(xeroTenantId, "string");
            localVarHeaderParams['Idempotency-Key'] = models_1.ObjectSerializer.serialize(idempotencyKey, "string");
            localVarHeaderParams['Accept'] = acceptHeadersFromSpec.join();
            Object.assign(localVarHeaderParams, options.headers);
            let localVarUseFormData = false;
            let localVarRequestOptions = {
                method: 'PUT',
                params: localVarQueryParameters,
                headers: localVarHeaderParams,
                url: localVarPath,
                responseType: responseTypeOption,
                data: models_1.ObjectSerializer.serialize(timesheetLine, "TimesheetLine"),
            };
            let authenticationPromise = Promise.resolve();
            authenticationPromise = authenticationPromise.then(() => this.authentications.OAuth2.applyToRequest(localVarRequestOptions));
            authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
            return authenticationPromise.then(() => {
                if (Object.keys(localVarFormParams).length) {
                    if (localVarUseFormData) {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'Content-Type': 'multipart/form-data' });
                    }
                    else {
                        localVarRequestOptions.data = localVarFormParams;
                        localVarRequestOptions.headers = Object.assign(Object.assign({}, localVarRequestOptions.headers), { 'content-type': 'application/x-www-form-urlencoded' });
                    }
                }
                return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
                    let body = null;
                    try {
                        const response = yield (0, axios_1.default)(localVarRequestOptions);
                        body = models_1.ObjectSerializer.deserialize(response.data, "TimesheetLineObject");
                        if (response.status && response.status >= 200 && response.status <= 299) {
                            resolve({ response: response, body: body });
                        }
                        else {
                            reject({ response: response, body: body });
                        }
                    }
                    catch (error) {
                        const errorResponse = new ApiError_1.ApiError(error);
                        reject(JSON.stringify(errorResponse.generateError()));
                    }
                }));
            });
        });
    }
}
exports.PayrollAuV2Api = PayrollAuV2Api;
