// xero-node/dist/gen/model/accounting/taxBreakdownComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaxBreakdownComponent = void 0;
class TaxBreakdownComponent {
    static getAttributeTypeMap() {
        return TaxBreakdownComponent.attributeTypeMap;
    }
}
exports.TaxBreakdownComponent = TaxBreakdownComponent;
TaxBreakdownComponent.discriminator = undefined;
TaxBreakdownComponent.attributeTypeMap = [
    {
        "name": "taxComponentId",
        "baseName": "TaxComponentId",
        "type": "string"
    },
    {
        "name": "type",
        "baseName": "Type",
        "type": "TaxBreakdownComponent.TypeEnum"
    },
    {
        "name": "name",
        "baseName": "Name",
        "type": "string"
    },
    {
        "name": "taxPercentage",
        "baseName": "TaxPercentage",
        "type": "number"
    },
    {
        "name": "taxAmount",
        "baseName": "TaxAmount",
        "type": "number"
    },
    {
        "name": "taxableAmount",
        "baseName": "TaxableAmount",
        "type": "number"
    },
    {
        "name": "nonTaxableAmount",
        "baseName": "NonTaxableAmount",
        "type": "number"
    },
    {
        "name": "exemptAmount",
        "baseName": "ExemptAmount",
        "type": "number"
    },
    {
        "name": "stateAssignedNo",
        "baseName": "StateAssignedNo",
        "type": "string"
    },
    {
        "name": "jurisdictionRegion",
        "baseName": "JurisdictionRegion",
        "type": "string"
    }
];
(function (TaxBreakdownComponent) {
    let TypeEnum;
    (function (TypeEnum) {
        TypeEnum[TypeEnum["USCOUNTRY"] = 'SYSGST/USCOUNTRY'] = "USCOUNTRY";
        TypeEnum[TypeEnum["USSTATE"] = 'SYSGST/USSTATE'] = "USSTATE";
        TypeEnum[TypeEnum["USCOUNTY"] = 'SYSGST/USCOUNTY'] = "USCOUNTY";
        TypeEnum[TypeEnum["USCITY"] = 'SYSGST/USCITY'] = "USCITY";
        TypeEnum[TypeEnum["USSPECIAL"] = 'SYSGST/USSPECIAL'] = "USSPECIAL";
    })(TypeEnum = TaxBreakdownComponent.TypeEnum || (TaxBreakdownComponent.TypeEnum = {}));
})(TaxBreakdownComponent || (exports.TaxBreakdownComponent = TaxBreakdownComponent = {}));
